helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -525,7 +526,7 @@ public void invoke(NotificationContext changeContext) throws Exception {
 
   private <T extends HelixProperty> List<T> preFetch(PropertyKey key) {
     if (_preFetchEnabled) {
-      return _accessor.getChildValues(key);
+      return _accessor.getChildValues(key, true);
```

Comment:
Same question. Before this change, partial read would not terminate rest flow, so onLiveInstance(), onMessage() etc would be called. 
Now if exception thrown, this is not the case. This would 
1/ result onLiveInstance() etc not called, but the exception got caught in the ZkEvent thread.
2/ For Molly's change to add perodical checking messages PULL request, this would break the time thread which calls onMessage(). Basically if the exception thrown from onMessage() when timer thread calls this one, this would stop later timer thread I believe. (We can test. I recall Java Concurrency Controller book mentioned this issue there.) This would defeat the purpose of Molly's change.

Reply:
"partial read would not terminate rest flow", true. But it will cause the controller to make the wrong decision. Which is even worse. To prevent this is the target of this change.

I don't think it will fail the ZkEvent thread. All the callers as I can find has been try-catched. Do you have an example of exception?

Reply:
Note, if exception thrown. For this CallbackHandler, before this change, partial read result would be feed to OnMessage() or OnWhatever; now, onMessage() would not be called, instead, zkEvent would get an exception, catch the exception nad do logging. This is what we want.

Then, in Molly's case, the timer thread can see exception too, right? The thing is that I suspect that timer's internal thread may not schedule more OnMessage() if the previous one has exception thrown.  This needs to be tested and verified. Otherwise, the timer thread may never work after exception thrown if my suspicion is right.

Reply:
@kaisun2000, that change has not been checked in yet. But it is a good point.
@mgao0, Please be aware of what Kai mentioned.

Reply:
Let me add this point to @mgao0's pull request.